### PR TITLE
feat: add test cases to existing test run via updateTestRun

### PR DESCRIPTION
**Pull request by SavioBS629**

### CODE REVIEW

File: src/tools/testmanagement-utils/update-testrun.ts [hunk 1]
```diff
@@ -6,6 +6,17 @@ import { formatAxiosError } from "../../lib/error.js";
 import { BrowserStackConfig } from "../../lib/types.js";
 import { getTMBaseURL } from "../../lib/tm-base-url.js";
 
+/**
+ * Selection of test cases (with optional configurations) to add.
+ */
+const TestCaseSelectionSchema = z.object({
+  test_case_ids: z.array(z.string()).describe("Test case IDs, e.g. TC-123"),
+  configuration_ids: z
```
Comment (SavioBS629) [inline src/tools/testmanagement-utils/update-testrun.ts:17]: **[Medium] No `.min(1)` on `test_case_ids` — empty selection fires a no-op PATCH reported as success**

`hasTestCases` only checks the outer `add_test_cases` array length, so `add_test_cases: [{ test_case_ids: [] }]` fires a real PATCH to `/test-cases` with an empty selection and returns `"...(added 0); changes apply asynchronously."` as if it did something — a real API side effect masquerading as a no-op.

**Suggestion:** Add `.min(1)` to `test_case_ids` (and consider it on `add_test_cases`), and/or filter empty selections before building the request body.

```suggestion
  test_case_ids: z
    .array(z.string())
    .min(1)
    .describe("Test case IDs, e.g. TC-123"),
```

_Reviewer: stack:code-reviewer_

File: src/tools/testmanagement-utils/update-testrun.ts [hunk 2]
```diff
@@ -85,3 +158,68 @@ export async function updateTestRun(
...
+    const { add_test_cases, preserve_existing_results } = args.test_run;
+
+    const body = {
+      test_run: {
+        add_test_cases,
+        preserve_existing_results: preserve_existing_results ?? true,
+      },
+    };
+
+    const tmBaseUrl = await getTMBaseURL(config);
```
Comment (SavioBS629) [inline src/tools/testmanagement-utils/update-testrun.ts]: **[Medium] Duplicate `getTMBaseURL` round-trips when both concerns are supplied**

Both `updateTestRunMetadata` and this helper call `getTMBaseURL(config)` independently. The module-level cache in `tm-base-url.ts` is intentionally skipped in `REMOTE_MCP` mode, so a call supplying both `name`/`run_state` and `add_test_cases` runs the region-discovery probe twice — doubling latency/API load, and letting the two sub-requests theoretically land on different regions if one probe flakes.

**Suggestion:** Resolve `tmBaseUrl` once in the `updateTestRun` dispatcher and pass it into both helpers; optionally `Promise.all` the two independent PATCHes.

_Reviewer: stack:code-reviewer_

**Review by SavioBS629 — COMMENTED** [review]
Claude Code Review (automated) — 2 inline finding(s). Full report in the PR comment below. Verdict: Passed.

**Comment by SavioBS629** [comment]
# Claude Code PR Review

**PR:** https://github.com/browserstack/mcp-server/pull/333  •  **Head:** `e7cfff0`  •  **Reviewers:** stack:code-reviewer

## Summary
Splits `updateTestRun` into a dispatcher plus two endpoint-specific helpers (`updateTestRunMetadata` → `.../update`, `updateTestRunTestCases` → `.../test-cases`) and adds an `add_test_cases` / `preserve_existing_results` capability, with a new 8-case vitest suite. Core dispatch, error propagation, and tests are correct; no security or data-corruption issues. A few efficiency/validation/coverage items are worth addressing before merge.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | Auth from config; Basic header matches existing pattern |
| High | Security | Authentication/authorization checks present | Pass | Same `getBrowserStackAuth` flow as sibling tools |
| High | Security | Input validation and sanitization | Pass | Zod schema; IDs url-encoded — but see empty-array gap below |
| High | Security | No IDOR — resource ownership validated | N/A | Delegated to BrowserStack API |
| High | Security | No SQL injection (parameterized queries) | N/A | No SQL |
| High | Correctness | Logic is correct, handles edge cases | Pass | Metadata path not regressed (base only ever had name/run_state); dispatch sound |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | `isError` OR'd across both concerns; failures never masked |
| High | Correctness | No race conditions or concurrency issues | Pass | Sequential; see perf note on parallelizing |
| Medium | Testing | New code has corresponding tests | Pass | Dedicated 187-line suite added |
| Medium | Testing | Error paths and edge cases tested | Fail | `catch`/rejection paths, `configuration_ids` forwarding, both-fail case uncovered |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Metadata behavior preserved; assertions match impl |
| Medium | Performance | No N+1 queries or unbounded data fetching | Fail | `getTMBaseURL` resolved twice per dual-concern call (uncached in remote mode) |
| Medium | Performance | Long-running tasks use background jobs | N/A | Backend applies test-case add asynchronously |
| Medium | Quality | Follows existing codebase patterns | Pass | Matches sibling tools |
| Medium | Quality | Changes are focused (single concern) | Pass | Focused feature + tests |
| Low | Quality | Meaningful names, no dead code | Pass | — |
| Low | Quality | Comments explain why, not what | Pass | Good dispatcher docstring |
| Low | Quality | No unnecessary dependencies added | Pass | — |

## Findings

- **File:** `src/tools/testmanagement-utils/update-testrun.ts:180` (and the metadata helper's own `getTMBaseURL` call)
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer
- **Issue:** `updateTestRunMetadata` and `updateTestRunTestCases` each call `getTMBaseURL(config)` independently. Per `src/lib/tm-base-url.ts`, the module-level cache is intentionally skipped in `REMOTE_MCP` mode (the deployment this reviewing repo targets), so a call that supplies both `name`/`run_state` and `add_test_cases` performs the region-discovery probe twice, doubling latency/API load and theoretically allowing the two sub-requests to land on different regions if one probe flakes.
- **Suggestion:** Resolve `tmBaseUrl` once in the dispatcher and pass it into both helpers; optionally `Promise.all` the two independent PATCHes.

- **File:** `src/tools/testmanagement-utils/update-testrun.ts:14`
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer
- **Issue:** Neither `TestCaseSelectionSchema.test_case_ids` nor the outer `add_test_cases` array has `.min(1)`. `hasTestCases` only checks the outer array length, so `add_test_cases: [{ test_case_ids: [] }]` fires a real PATCH to `/test-cases` with an empty selection and reports `"...(added 0); changes apply asynchronously."` as if it did something — a real API side effect masquerading as a no-op.
- **Suggestion:** Add `.min(1)` to `test_case_ids` (and consider it for `add_test_cases`), and/or filter empty selections before computing `hasTestCases`/building the body.

- **File:** `src/tools/testmanagement-utils/update-testrun.ts:122`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** The `Authorization: "Basic " + Buffer.from(...)` block is now duplicated verbatim in both helpers (was single before the split).
- **Suggestion:** Extract a `buildAuthHeader(config)` helper shared with the existing pattern.

- **File:** `tests/tools/update-testrun.test.ts` (whole file)
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** Coverage gaps: neither `catch`/`formatAxiosError` branch is exercised (no rejected `apiClient.patch`); `configuration_ids` forwarding is never asserted; only "one call fails" is tested (not "both fail" surviving the `flatMap`); no empty-`test_case_ids` test.
- **Suggestion:** Add rejection-path tests for both helpers, a `configuration_ids` forwarding assertion, a both-fail case, and (after the `.min(1)` fix) an empty-selection rejection test.

### Non-blocking notes
- **[Info]** When both concerns run and one fails, the aggregated `content` blocks are concatenated unlabeled — a caller infers which failed from wording only. Consider prefixing blocks (`[metadata]` / `[test-cases]`) or a short summary line.
- **[Info, pre-existing]** `README.md:337` still describes `updateTestRun` as covering "tags, notes" (never implemented) and omits the new `add_test_cases` capability — worth refreshing here since it's on-topic.

### Raised by other reviewers (not independently confirmed)
_None — no human reviewer comments on this PR._

---

**Verdict: PASS** — the refactor and new capability are correct with a solid test suite and no security/regression issues; the two Medium items (double base-URL resolution in remote mode, empty-selection validation gap) and the Low test/dedup items are recommended before merge but non-gating.

File: src/tools/testmanagement-utils/update-testrun.ts [hunk 3]
```diff
@@ -27,33 +41,104 @@ export const UpdateTestRunSchema = z.object({
...
 });
 
 type UpdateTestRunArgs = z.infer<typeof UpdateTestRunSchema>;
 
+/**
+ * Builds the HTTP Basic auth header from per-request config credentials.
+ */
+function buildAuthHeader(config: BrowserStackConfig): string {
+  const [username, password] = getBrowserStackAuth(config).split(":");
+  return "Basic " + Buffer.from(`${username}:${password}`).toString("base64");
```
Comment (SavioBS629) [inline src/tools/testmanagement-utils/update-testrun.ts]: **[Low] Redundant split/rejoin in `buildAuthHeader`**

`getBrowserStackAuth(config)` already returns exactly `username:accesskey` — the string to base64-encode. Splitting on `:` and re-joining is redundant, and if an access key ever contained a `:`, the destructure would truncate the credential and yield a silent 401.

**Suggestion:**
```suggestion
  return "Basic " + Buffer.from(getBrowserStackAuth(config)).toString("base64");
```

_Reviewer: stack:code-review_

**Review by SavioBS629 — COMMENTED** [review]
Claude Code Review (automated) — 1 inline finding(s). Full report in the PR comment below. Verdict: Passed.

**Comment by SavioBS629** [comment]
# Claude Code PR Review

_Continues [the previous review](https://github.com/browserstack/mcp-server/pull/333#issuecomment-4914042194) — changes since `e7cfff0` (delta re-review)._

**PR:** https://github.com/browserstack/mcp-server/pull/333  •  **Head:** f2680df  •  **Reviewers:** stack:code-review

## Summary
Refactors `updateTestRun` to address the prior review: hoists the `getTMBaseURL` region lookup and auth-header construction to a single call each, adds `.min(1)` on `test_case_ids`, and adds tests for the catch path, `configuration_ids` forwarding, empty-selection guard, and the both-concerns-fail case.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | Auth via `getBrowserStackAuth(config)`; no creds in schema or logs. |
| High | Security | Authentication/authorization checks present | Pass | Per-request `config` → `buildAuthHeader`; no `process.env` reads. |
| High | Security | Input validation and sanitization | Pass | Zod schema; `.min(1)` added; IDs `encodeURIComponent`-escaped in URLs. |
| High | Security | No IDOR — resource ownership validated | N/A | Ownership enforced server-side by BrowserStack TM API. |
| High | Security | No SQL injection (parameterized queries) | N/A | No SQL; HTTP calls via `apiClient`. |
| High | Correctness | Logic is correct, handles edge cases | Pass | `Nothing to update` guard; empty inner/outer arrays handled. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Both helpers try/catch → `CallToolResult`; `Promise.all` cannot reject. |
| High | Correctness | No race conditions or concurrency issues | Pass | Two independent PATCHes; `Promise.all` order deterministic. |
| Medium | Testing | New code has corresponding tests | Pass | 4 new tests cover the changed paths. |
| Medium | Testing | Error paths and edge cases tested | Pass | Catch path, both-fail, empty-selection all covered. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Signature changes are internal; suite updated accordingly. |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | `getTMBaseURL` now called once per invocation (prior finding fixed). |
| Medium | Performance | Long-running tasks use background jobs | N/A | Test-case add is async server-side; response notes this. |
| Medium | Quality | Follows existing codebase patterns | Pass | Two-endpoint dispatch, util/registration split respected. |
| Medium | Quality | Changes are focused (single concern) | Pass | Scoped refactor addressing prior review. |
| Low | Quality | Meaningful names, no dead code | Pass | `buildAuthHeader`, `addIds` clear. |
| Low | Quality | Comments explain why, not what | Pass | Endpoint-split rationale documented. |
| Low | Quality | No unnecessary dependencies added | Pass | None added. |

## Findings

**Resolved since last review:**
- ~~`src/tools/testmanagement-utils/update-testrun.ts:180` Medium — Duplicate getTMBaseURL round-trips when both concerns supplied~~ Resolved — `getTMBaseURL` now called once and passed into both helpers (f2680df).
- ~~`src/tools/testmanagement-utils/update-testrun.ts:14` Medium — No .min(1) on test_case_ids; empty selection fires a no-op PATCH reported as success~~ Resolved — `.min(1)` added and `Nothing to update` guard gates empty selections (test added).
- ~~`src/tools/testmanagement-utils/update-testrun.ts:122` Low — Duplicated auth-header construction across both helpers~~ Resolved — extracted to `buildAuthHeader`.
- ~~`tests/tools/update-testrun.test.ts` Low — Test gaps: catch/exception paths, configuration_ids forwarding, both-fail case~~ Resolved — all four gaps now covered.

**Open (new this run):**

- **File:** `src/tools/testmanagement-utils/update-testrun.ts:61`
- **Severity:** Low
- **Reviewer:** stack:code-review
- **Issue:** `buildAuthHeader` does `getBrowserStackAuth(config).split(":")` then re-joins `${username}:${password}` before base64-encoding. The split/rejoin is redundant — `getBrowserStackAuth` already returns exactly `username:accesskey` — and if an access key ever contained a `:`, `split(":")` destructuring would truncate the credential and produce a silent 401. Very low likelihood given key format; mirrors a pre-existing pattern in `tm-base-url.ts`.
- **Suggestion:** `return "Basic " + Buffer.from(getBrowserStackAuth(config)).toString("base64");` — no split needed. Optionally hoist into a shared lib helper to retire the duplicate in `tm-base-url.ts`.

---

**Verdict: PASS** — the delta resolves all four prior findings; the one new item is a non-blocking Low.

**Comment by SavioBS629** [comment]
# Claude Code PR Review

_Continues [the previous review](https://github.com/browserstack/mcp-server/pull/333#issuecomment-4922668672) — changes since `f2680df` (delta re-review)._

**PR:** https://github.com/browserstack/mcp-server/pull/333  •  **Head:** ec43108  •  **Reviewers:** stack:code-review

## Summary
Single-commit follow-up (`ec43108 refactor: simplify buildAuthHeader`) that drops the redundant `split(":")`/rejoin in `buildAuthHeader` and base64-encodes the auth string directly — implementing the fix suggested in the previous review.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | Auth via `getBrowserStackAuth(config)`; encoded directly, matches `security.md` canonical pattern. |
| High | Security | Authentication/authorization checks present | Pass | Per-request `config`; no `process.env` reads. |
| High | Security | Input validation and sanitization | Pass | Unchanged by this delta; Zod schema intact. |
| High | Security | No IDOR — resource ownership validated | N/A | Enforced server-side by BrowserStack TM API. |
| High | Security | No SQL injection (parameterized queries) | N/A | No SQL. |
| High | Correctness | Logic is correct, handles edge cases | Pass | `Buffer.from("user:key")` is equivalent to the prior split/rejoin for well-formed creds, and no longer truncates keys containing `:`. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Unchanged. |
| High | Correctness | No race conditions or concurrency issues | Pass | Unchanged. |
| Medium | Testing | New code has corresponding tests | Pass | Existing auth-header assertions still cover the helper; behavior unchanged for standard creds. |
| Medium | Testing | Error paths and edge cases tested | Pass | Unchanged from prior delta. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Output byte-identical for standard `user:key` credentials. |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | Unchanged. |
| Medium | Performance | Long-running tasks use background jobs | N/A | — |
| Medium | Quality | Follows existing codebase patterns | Pass | Now matches the canonical `Basic + Buffer.from(authString)` pattern. |
| Medium | Quality | Changes are focused (single concern) | Pass | One 2-line simplification. |
| Low | Quality | Meaningful names, no dead code | Pass | Removed the unused destructure. |
| Low | Quality | Comments explain why, not what | Pass | Doc comment retained and accurate. |
| Low | Quality | No unnecessary dependencies added | Pass | None. |

## Findings

**Resolved since last review:**
- ~~`src/tools/testmanagement-utils/update-testrun.ts:61` Low — `buildAuthHeader` splits then rejoins the auth string; redundant, truncates keys containing `:`~~ Resolved — `ec43108` encodes `getBrowserStackAuth(config)` directly, matching the `security.md` canonical pattern.

No open findings.

---

**Verdict: PASS** — the follow-up resolves the only open finding; no new issues.

**Review by ruturaj-browserstack — APPROVED** [review]
